### PR TITLE
Components improve

### PR DESCRIPTION
**Pull request by Dantara**
**Improved version of connectivity components find.** 
Changelog:
1. Replace `vertexList` with self-made `extractVertices` function. It can have duplicates and output is not sorted but it has better time complexity.
2. `DisjointSet` has `Monoid` instance. Functionality in "overlay" section of `component` function can be replaced with `mappend`:
`components (Overlay x y) = (components x) <> (components y)`

### CODE REVIEW

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 1]
```diff
@@ -1,30 +1,38 @@
 module Algebra.Graph.Algorithm.Internal where
 
-import           Algebra.Graph
+import                            Algebra.Graph
+import Data.DisjointSet           (DisjointSet)
```
Comment (unaimillan) [inline src/Algebra/Graph/Algorithm/Internal.hs:4]: Is there any reason to import `DisjointSet` specifically instead of using `DS.DisjointSet`?

Comment (Dantara) [reply src/Algebra/Graph/Algorithm/Internal.hs:4]: Remove `DS.` prefix in type declarations:
`components :: Ord a => Graph a -> DisjointSet a`

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 2]
```diff
@@ -1,30 +1,38 @@
...
+-- "{{1,2},{3,4}}"
+--
+-- >>> DS.pretty $ components (((1 * 2) + (3 * 4)) + (2 * 3))
+-- "{{1,2,3,4}}"
+components :: Ord a => Graph a -> DisjointSet a
 components Empty         = DS.empty
 components (Vertex x)    = DS.singleton x
-components (Overlay x y) = mergeComponents (components x) (components y)
-components (Connect x y) = DS.fromLists [vertexList x ++ vertexList y]
+components (Connect x y) = DS.fromLists [extractVertices x <> extractVertices y]
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:17]: Why not use `vertexSet`?

```suggestion
components g@(Connect _ _) = DS.singletons (vertexSet g)
```

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 3]
```diff
@@ -1,30 +1,38 @@
...
+components (Overlay x y) = (components x) <> (components y)
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:18]:
```suggestion
components (Overlay x y) = components x <> components y
```

This relies on `DS.append` (which is not exported for some reason) which seems to be O(m log n), m <= n.
So overall complexity of `components` right now seems to be O(n log^2 n).

I think it should be possible to provide a more efficient version of `DS.append` similar to `Map.union`, which should eliminate one extra _log n_.

File: src/Algebra/Graph/Algorithm/Internal.hs (hunk 1 above)
Comment (fizruk) [reply src/Algebra/Graph/Algorithm/Internal.hs:4]: This is common import practice, similar to

```haskell
import Data.IntMap (IntMap)
import qualified Data.IntMap as IntMap
```

This is recommended in the [documentation of `containers`](https://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-IntMap.html), for instance.

One comment would be that `DS` is not worth abbreviation, so I would go with

```haskell
import Data.DisjointSet (DisjointSet)
import qualified Data.DisjointSet as DisjointSet
```

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 4]
```diff
@@ -1,30 +1,18 @@
...
+-- >>> DisjointSet.pretty $ components ((1 * 2) + (3 * 4))
+-- "{{1,2},{3,4}}"
+--
+-- >>> DisjointSet.pretty $ components (((1 * 2) + (3 * 4)) + (2 * 3))
+-- "{{1,2,3,4}}"
+components :: Ord a => Graph a -> DisjointSet a
+components Empty           = DisjointSet.empty
+components (Vertex x)      = DisjointSet.singleton x
+components g@(Connect _ _) = DisjointSet.singletons $ vertexSet g
+components (Overlay x y)   = components x <> components y
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:18]: One idea that will rely on a _mutable_ disjoint set, which has good known implementations:

```haskell
components :: Ord a => Graph a -> DisjointSet a
components g = runST $ do
  mds <- initEmptyMDisjointSet ...
  componentsM g mds
  freeze mds

componentsM :: Ord a => Graph a -> MDisjointSet s a -> ST s ()
componentsM = ...
```

Comment (fizruk) [reply src/Algebra/Graph/Algorithm/Internal.hs:18]: Looks like we could use `ST`-based implementation from here: https://hackage.haskell.org/package/union-find-0.2/docs/Data-UnionFind-ST.html

**Pull request merged**